Revenant: A Mini-Antagonist

Xhuis opened the pull request:
This is a work-in-progress miniature antagonist. By 'miniature antagonist' I mean it is spawned by event in the middle of an existing round and is not determined at the beginning of the round, similar to the space ninja.

The revenant is a unique foe. It is similar to a ghost - it is invisible to the naked eye, can phase through walls, and is invincible while hidden. It cannot talk to normal ghosts - it can't talk at all, actually - and lacks typical ghost powers such as teleportation. The revenant uses a unique resource called Essence, or E for short, that regenerates in small amounts on its own. However, in order to use more powerful abilities, the revenant must siphon essence from dead or dying humans. It will be vulnerable during this time. The max health of a revenant is twice the amount of essence it has; if a wraith has 25E, its max health is 50.

Picture of a revenant manifesting to 'harvest' a dying human:
![harvest1](https://cloud.githubusercontent.com/assets/5659269/7058951/ad5d9096-de36-11e4-80e9-eb61e9286f82.gif)

The essence gained from Harvest depends on the state of the human. Previously-intelligent humans will yield more essence, and unconscious humans will yield much more essence than dead ones.

The revenant has many abilities at its disposal - however, it must spend essence in order to unlock them. One of these abilities sends a silent message to a single mob - this is the wraith's only way of directly communicating. More powerful abilities require more essence, and essence can be obtained in greater quantities the more powerful the revenant gets.

While revenants are invisible to normal enemies, machinery like turrets, as well as unintelligent creatures, will be able to sense the revenant.

One weakness of the revenant is holy items. The null rod and bible are both effective in dealing with a revenant, and a few whacks from either can do substantial amounts of damage due to their essence-draining powers. Revenants can only reach minimum essence so many times before they die for good.

**_Revenants do not have night vision.**_

**Ability List**

_Lesser Harvest (10 second cooldown, 0E)_ will target a corpse or unconscious human and drain a small amount of essence from them. It is the only spell unlocked by default and is necessary for the revenant to begin. While draining, you will be exposed for a decent amount of time.

_Greater Harvest (2 minute cooldown, 125E unlock cost, 0E use cost)_ is a much more powerful version of its lesser counterpart. This ability will generate more than twice the essence as before and can also target living humans to kill them. It is a good idea to greater drain someone to death and then lesser drain them afterwards.

_Transmit (5 second cooldown, 5E unlock/use cost)_ is the wraith's only direct way of communication. For a small amount of essence, the wraith can send a single message to a single target in distinct text. Strike deals with antagonists, or tell people spooky things.

_Fling Objects (20 second cooldown, 20E unlock/use cost)_ will pick up all nearby objects and throw them at the target human. This includes any item, such as toolboxes, radio headsets, backpacks, you name it. Use near broken windows for FUN! Note: the revenant is also exposed for about two seconds while using this ability.

_Overload Light (30 second cooldown, 25E unlock/use cost)_ is a very unique ability. All nearby light fixtures that are toggled on will be overloaded with essence, and, after a short time, send out bolts of electricity to shock everyone nearby for a decent amount of damage.

_Strangulate (50 second cooldown, 30E unlock/use cost)_ will deal a high amount of oxygen damage to a target in addition to quite some time of being unable to breathe or speak. After a short time, if they remain outside of critical condition, the oxygen damage will instantly be healed. Ideally, this can be used in conjunction with other abilities.

_Life Tap (3 second cooldown, 25E unlock cost, 0E use cost)_ will drain a single strike to grant a hefty sum of 50E. Strikes are used to prevent death if your essence reaches zero, and you start with three by default.

_Seed of Draining (2 minute cooldown, 20E unlock/use cost)_ will corrupt a living human with draining energies. For the next two minutes, the target will take continuous oxygen damage and the revenant will gain one essence point. This is the primary way of gaining essence both nonlethally and after Lesser Harvest becomes unfeasible.

dannno commented:
child rape

SconesC commented:
So sort of like Lifeweb?

On that note, please for the love of god do not add a possession power

Supermichael777 commented:
you played lifeweb? concept seems similar but less annoyingly location dependent.

Xhuis commented:
I have no idea what Life web is.

dannno commented:
haha ok buddy

Czarcastic commented:
Gotta name it something else or it might get confused with construct wraiths

Steelpoint commented:
Reminds me somewhat of that unique ghost antagonist on Goonstation. I forget its name but it functions somewhat the same. 

I also do recall that lines of salt or suger will reveal the Goon ghost.

Either way this seems like a very unique antagonist, so :+1:

Alek2ander commented:
>I also do recall that lines of salt or suger will reveal the Goon ghost.

So that's why chempiles were added. It all makes sense now.

KorPhaeron commented:
Seconding Czar. I suggest Revenant.

Cool stuff though.

AlexCrimson commented:
So whats the point? Does it eventually gain the ability to kill people? Does it have objectives? Is there any reason for the crew to care?

Xhuis commented:
@AlexCrimson: The more essence a wraith absorbs, the more powerful it grows. Its goal is to gain as much essence as possible using its abilities. It does gain the ability to do direct damage, and it will start killing people off if the crew don't do anything about it.

I plan on making tiles with holy water on them reveal wraiths. In addition, floorhitting with a bible will reveal and stun for a brief period.

Steelpoint commented:
One reason why I like the salt lines revealing Wraiths is that Salt is very visible, whereas Holy Waterd tiles are not so visable.

Stickymayhem commented:
This is awesome. Make sure the null rod is extremely effective on it too.

I'd love to see more weird alien things ending up on station and being minor nuisances. It immediately adds more things to mix into a round for a more unique shift every time.

Xhuis commented:
I'm gonna go ahead and say for now that this is done. It was a miniature project, and it's pretty small. Wraiths have been renamed to Revenants, there is an event which spawns them in, they have new abilities, and chaplain stuff is powerful against them. I'll do some updating on the OP right now.

RemieRichards commented on code/modules/mob/living/simple_animal/revenant/revenant.dm:
```diff
@@ -0,0 +1,354 @@
+			if(istype(O, /obj/item/device/radio/intercom))
+				continue
+			objects_to_throw.Add(O)
+		for(var/obj/item/O in orange(3, M))
```
surely just loop the ```objects_to_throw``` list?

Also, consider ```spawn()```ing this and some of the other effects off, so that they happen simultaneously instead of one by one.

Xhuis commented:
I'll see what I can do. Also, I never actually learned, what is the difference between spawn() and sleep()?

RemieRichards commented:
@Xhuis Sleep waits the current proc for the time specified.
Spawn prolongs the proc for the time specified, allowing other things to happen at the same time.

Xhuis commented:
So would putting a spawn(0) before the actual code in functions be enough to accomplish that effect?

RemieRichards commented:
Look at how I throw objects during explosions, It's spawn(0) at the beginning of each loop iteration.

Stickymayhem commented:
So where can we look up the abilities. Is a wiki page coming?

AnturK commented on code/__DEFINES/preferences.dm:
```diff
@@ -40,3 +40,4 @@
 #define BE_NINJA		1024
 #define BE_MONKEY		2048
 #define BE_GANG			4096
+#define BE_REVENANT		8192 //This is gonna conflict with shadowling :/
```
With this, shadowling and abduction we're running out of bits. So some refactor will be needed.

AnturK commented:
Default event weight seems bit much for something that's very hard to kill annoyance. Then again there's not enough events that have impact on the round so whatever.

KorPhaeron commented:
There really arent. Most of the anomalies dont seem to do anything

Cheridan commented:
anomalies are awesome nerd i'll cut you

d3athrow commented:
blatant theft from goon wow

Dorsisdwarf commented:
Sprite should be animated to look like a ghost instead of a vinyl cutout.
So should the 'appeaaring' be animated.
Also it would be nice if it bobbed like actual ghosts.

Xhuis commented:
@Dorsisdwarf: It does bob like ghosts while it isn't draining. The other sprites I'm happy with already.

Xhuis commented:
Will fix conflicts when I can be bothered

Durkel commented:
If it hasn't been added or suggested could you make it so it can't pass over blessed floor tiles?

Xhuis commented:
Would probably be too hard to code currently, but I'll see what I can do.

Antimattercarp commented:
Freaking awesome :+1:x:100:

KorPhaeron commented:
Take a look at the jaunt code if you wanr holy tiles to stop it.

It'd make it very easy to kill though considering how an extinguisher+tank of holy water quickly covers the station

JJRcop commented:
Actually not that hard to code it. Just override Move() and return 0 if the turf has holy water. If you want to be cheaper you can just check if the turf has the NOJAUNT flag set.

Xhuis commented:
I'll get around to working on this once I finish up changes on shadowling

Commit:
Revenants

Xhuis commented:
Merge-ready. Wiki page is also created at https://tgstation13.org/wiki/Revenant

Xhuis commented:
Removed the murderbone abilities and renamed Lesser Harvest to Harvest. More abilities coming soon.

AlexCrimson commented:
Who bitched at you? I imagine someone did. I was looking forward to force-choking someone too.

GunHog commented:
Nerfing it even before it is merged? Geez.

Xhuis commented:
On GitHub someone made a good point that you could start murderboning for infinity like three minutes after spawning. I want to go with a more nonlethal approach now.

AlexCrimson commented:
Then whats the point? If its a ghost that cannot hurt you then there is no point trying to kill it or disrupt its objectives. No point calling it an antagonist if it cannot antagonize anyone.

boring/10

Xhuis commented:
Squashed commits and added two lethal abilities: Mind Spike and Mind Blast. They have a custom sound, which can be listened to here.
https://dl.dropboxusercontent.com/u/116194515/mind_blast.ogg

PKPenguin321 commented:
update the wiki page for mind spike/blast pls

Miauw62 commented:
I think BE_REVENANT will not work, you can only do bitwise operations up to 16 bits.

Miauw62 commented on code/modules/mob/living/simple_animal/revenant/revenant.dm:
```diff
@@ -0,0 +1,134 @@
+			strikes--
+			src << "<span class='boldannounce'>Your essence has dropped below critical levels. You barely manage to save yourself - [strikes ? "you can't keep this up!" : "next time, it's death."]</span>"
+		else if(strikes <= 0)
+			src << "<span class='warning'><b>NO! No... it's too late, you can feel yourself fading...</b></span>"
```
use danger or userdanger.

Miauw62 commented on code/modules/mob/living/simple_animal/revenant/revenant.dm:
```diff
@@ -0,0 +1,134 @@
+	maxHealth = essence * 2
+	if(!revealed)
+		health = maxHealth //Heals to full when not revealed
+	if(essence < essence_regen_cap && essence_regen)
```
this check would probably best be
```if(essence_regen)
    essence = max(essence_regen_cap, essence + 1)
```

Jalleo commented on code/modules/mob/living/simple_animal/revenant/revenant.dm:
```diff
@@ -0,0 +1,134 @@
+	if(istype(W, /obj/item/weapon/nullrod))
+		src.visible_message("<b>The revenant</b> screeches and flails!", \
+							"<span class='userdanger'>The null rod invokes agony in you! You feel your essence draining away!</span>")
+		src.essence -= 25 //hella effective
```
Could that make the essence go negative?

Miauw62 commented on code/modules/mob/living/simple_animal/revenant/revenant.dm:
```diff
@@ -0,0 +1,134 @@
+
+/mob/living/simple_animal/revenant/proc/change_essence_amount(var/essence_amt, var/mode = 0, var/silent = 0, var/source = null, var/mob/living/simple_animal/revenant/user = usr)
+	//Mode 1 is essence subtracted, mode 0 is essence added
+	//Example use: revenant.change_essence_amount(25, 0, 0, "the debug") would tell him "Gained 25E from the debug."
```
Can't you just leave out the mode argument and call this with negative amounts of essence?
e.g. change_essence_amount(-25, 0, "the debug") instead of change_essence_amount(25, 1, 0, "the debug")

Miauw62 commented on code/modules/mob/living/simple_animal/revenant/revenant.dm:
```diff
@@ -0,0 +1,134 @@
+		src.visible_message("<b>The revenant</b> screeches and flails!", \
+							"<span class='userdanger'>The null rod invokes agony in you! You feel your essence draining away!</span>")
+		src.essence -= 25 //hella effective
+		if(prob(5))
```
This is pretty lame. If the essence runs out the revenant is going to lose a strike anyway.

Miauw62 commented on code/modules/mob/living/simple_animal/revenant/revenant_abilities.dm:
```diff
@@ -0,0 +1,318 @@
+				user << "<span class='warning'>You are already siphoning the essence of a soul!</span>"
+				return
+			if(target in drained_mobs)
+				user << "<span class='warning'>[target]'s soul is dead and empty.</span>"
```
Doesn't this mean that multiple revenants will be able to harvest the same person?

Xhuis replied:
Yes, it does. It is for balance reasons that every revenant can only harvest one time so their essence doesn't skyrocket, but it's also intended for new revenants to be able to get essence from other drained corpses too.

Miauw62 commented on code/modules/mob/living/simple_animal/revenant/revenant_abilities.dm:
```diff
@@ -0,0 +1,318 @@
+		return
+	for(var/mob/living/M in targets)
+		spawn(0)
+			var/msg = stripped_input(usr, "What do you wish to tell \the [M.name]?.", null, "")
```
You can replace ```\the [M.name]``` with ```[M]```, it will yield the same result.

Miauw62 commented on code/modules/mob/living/simple_animal/revenant/revenant_abilities.dm:
```diff
@@ -0,0 +1,318 @@
+//Life Tap: Drains one 'strike' to gain 50E
+/obj/effect/proc_holder/spell/targeted/revenant_life_tap
+	name = "Unlock: Life Tap (25E)"
+	desc = "Draws from your own life tool to gain more essence. Can only be cast three times."
```
life tool?

Cheridan replied:
it's a euphemism for penis, right?

AlexCrimson commented:
Re-add choke but have it deal a much smaller amount of oxygen damage. So its more of an ability that pushes people over the edge rather than instantly kills them. Make it something you unlock near the endgame, so only when people know about you. The idea is that it punishes people who run around heavily damaged whilst they know the Revenant is nearby. Would also need to give it a big cooldown so you cannot spam it against one target.

I think thats fair, no?

Could you add an ability that "curses" an area? Lowering the temperature a lot? Not sure if thats possible. Some sort of freeze attack seems appropriate for a ghost.

Miauw62 commented on code/modules/mob/living/simple_animal/revenant/revenant_abilities.dm:
```diff
@@ -0,0 +1,318 @@
+			sleep(10)
+			user.essence++ //Drains 120 essence total
+			if(prob(1))
+				target << "You feel a curious draining sensation."
```
120 essence is quite a lot.
this should probably induce some staminaloss over time instead of a dumb 1% chance of the equivalent of "you feel a tiny prick!"

Miauw62 commented:
none of these abilities (not even the direct damage abilities) seem to reveal the revenant?

Xhuis commented:
Real life is catching up with me so I'm going to close this until I get more of a chance to work on.

__Revenant is not dead, just on hiatus.__

AlexCrimson commented:
Technically it was always dead...

...hehe

Sorry.

Commit:
More progress

Xhuis commented:
Yeah whatever, fuck it.

Commit:
Fixes two oversights

AlexCrimson commented:
but... my force choke =(

Commit:
Fixed some more issues

Miauw62 commented:
oh woop i didnt notice you fixed my issues. comment next time, i dont get mails about commits.

PKPenguin321 commented:
:ghost:

Jordie0608 commented:
code/modules/mob/living/simple_animal/revenant/revenant_abilities.dm:238:warning: M: variable defined but not used

AlexCrimson commented:
Tried it on a test server using the current github version. Ran into a few issues after spawning in a Revenant simple mob and possessing it.

The sound for Mind Spike/Blast is way too obnoxious. Should be toned down a lot.

When attacked by the abilities i got this message "<span class'boldannounce'>You feel a spike of pain in your head!"

The Revenants essence does not stop regenerating. I am currently at over 500E and still going up. No idea if this is because its an admin spawned Revenant and not an event one.

I didnt notice any Oxygen damage from Seed of Draining. Unless its so minor that it instantly gets healed.

Feels like the Revenant isnt visible for long enough to attack it. Harvest happens pretty quickly.

...Also pls re add the Telekinesis and Strangulate in some form. Because right now the abilities arnt much fun. Not asking for instant crit abilities.

Xhuis commented:
I'll make a quick update PR.

AlexCrimson commented:
Uh i copy/pasted the mind spike/blast thing wrong. It had some weird code message in the attack log for the human i hit with the ability...

<span class'boldannounce'>You feel a spike of pain in your head!

Is what it said.

Xhuis commented:
Oh, was there <span='boldannounce'> before it? I think that might be it.

AlexCrimson commented:
The angle brackets make it disappear when i post the comment. I have no clue why. Some github feature i imagine.

Xhuis commented:
Did it say something like "span class='boldannounce'" before the message?

AlexCrimson commented:
http://i.imgur.com/hMEzIha.png

There you go. That help?

Xhuis commented:
Yep, will be fixed soon. Thanks for the report.
